utils/inference_utils.py

```diff
@@ -844,19 +844,35 @@ def run_inference(examples, serving_bundle):
     return (common_utils.convert_prediction_values(values, serving_bundle),
             None)
   elif serving_bundle.custom_predict_fn:
-    # If custom_predict_fn is provided, pass examples directly for local
-    # inference.
-    values = serving_bundle.custom_predict_fn(examples)
-    extra_results = None
-    # If the custom prediction function returned a dict, then parse out the
-    # prediction scores. If it is just a list, then the results are the
-    # prediction results without attributions or other data.
-    if isinstance(values, dict):
-      preds = values.pop('predictions')
-      extra_results = values
-    else:
-      preds = values
-    return (common_utils.convert_prediction_values(preds, serving_bundle),
-            extra_results)
+    try:
+      # If custom_predict_fn is provided, pass examples directly for local
+      # inference.
+      values = serving_bundle.custom_predict_fn(examples)
+      extra_results = None
+      # If the custom prediction function returned a dict, then parse out the
+      # prediction scores. If it is just a list, then the results are the
+      # prediction results without attributions or other data.
+      if isinstance(values, dict):
+        preds = values.pop('predictions')
+        extra_results = values
+      else:
+        preds = values
+      return (common_utils.convert_prediction_values(preds, serving_bundle),
+              extra_results)
+
+    # The custom_predict_fn for colab/jupyter accepts one parameter.
+    # While the custom_predict_fn for non-colab use have two.
+    # If an user wrongly write a custom_predict_fn with only one parameter,
+    # chances are that the `try` block will fail.
+    except Exception as e:
+      try:
+        return (serving_bundle.custom_predict_fn(examples, serving_bundle), None)
```

Comment:
why providing the serving bundle as a second arg? shouldn't we be able to use the same logic above for this custom case and not have this extra block?

Reply:
If the inference needs to compose URI from the address, model name etc, the information in  serving_bundle is needed. The extra block is a fall back case so that the original code path (custom function supplied from colab/jupyter) is intact.

Here's the sample code that uses serving_bundle:
https://github.com/pytorch/serve/pull/418/files#diff-d898e6bcafa81d84ef29e0fcee70bd21

Reply:
Ah I understand now. Instead of try/except here, how about checking the number of arguments in the custom_predict_fn and then calling it as necessary, instead of trying the first way and falling back if it fails (see https://stackoverflow.com/questions/847936/how-can-i-find-the-number-of-arguments-of-a-python-function and note that we only need to support py3 going forward)

Reply:
Thanks! This really makes code path clearer. Fixed.